src/client/gameui.cpp

```diff
@@ -255,28 +254,42 @@ void GameUI::updateChatSize()
 
 void GameUI::updateProfiler()
 {
-	if (m_profiler_current_page != 0) {
-		std::ostringstream os(std::ios_base::binary);
-		os << "   Profiler page " << (int)m_profiler_current_page <<
-				", elapsed: " << g_profiler->getElapsedMs() << " ms)" << std::endl;
-
-		g_profiler->print(os, m_profiler_current_page, m_profiler_max_page);
+	m_guitext_profiler->setVisible(m_profiler_current_page != 0);
+	if (m_profiler_current_page == 0)
+		return;
 
-		EnrichedString str(utf8_to_wide(os.str()));
-		str.setBackground(video::SColor(120, 0, 0, 0));
-		setStaticText(m_guitext_profiler, str);
+	std::ostringstream oss(std::ios_base::binary);
+	oss << "(Profiler page " << (int)m_profiler_current_page
+		<< "/" << (int)m_profiler_max_page
+		<< ", elapsed: " << g_profiler->getElapsedMs() << " ms)" << std::endl;
+	g_profiler->print(oss, m_profiler_current_page, m_profiler_max_page);
 
-		core::dimension2d<u32> size = m_guitext_profiler->getOverrideFont()->
-				getDimension(str.c_str());
-		core::position2di upper_left(6, m_guitext->getTextHeight() * 2.5f);
-		core::position2di lower_right = upper_left;
-		lower_right.X += size.Width + 10;
-		lower_right.Y += size.Height;
+	EnrichedString str(utf8_to_wide(oss.str()));
+	str.setBackground(video::SColor(120, 0, 0, 0));
+	setStaticText(m_guitext_profiler, str);
 
-		m_guitext_profiler->setRelativePosition(core::rect<s32>(upper_left, lower_right));
+	v2s32 upper_left(5, 10);
+	if (m_flags.show_minimal_debug)
+		upper_left.Y += m_guitext->getTextHeight();
+	if (m_flags.show_basic_debug)
+		upper_left.Y += m_guitext2->getTextHeight();
+
+	v2s32 lower_right = upper_left;
+	lower_right.X += m_guitext_profiler->getTextWidth() + 5;
+	lower_right.Y += m_guitext_profiler->getTextHeight();
+
+	m_guitext_profiler->setRelativePosition(core::recti(upper_left, lower_right));
+
+	// Really dumb heuristic (we have a fixed number of pages, not a fixed page size)
+	const v2u32 window_size = RenderingEngine::getWindowSize();
+	if (m_guitext_profiler->getRelativePosition().UpperLeftCorner.Y + m_guitext_profiler->getTextHeight()
+		> window_size.Y * 0.7f) {
+		int max_page = std::min<u8>(5, m_profiler_max_page + 1);
```

Comment:
Why 5? What's the origin?

Reply:
It is, but only up to 5 pages.
The idea is that too many pages are too annoying to skip through. And also a windows height of <400 isn't realistic.